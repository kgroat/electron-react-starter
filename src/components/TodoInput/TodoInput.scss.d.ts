export const todoInput: string;
export const title: string;
export const titleDefault: string;
export const submitButton: string;
export const icon: string;
export const disabled: string;
