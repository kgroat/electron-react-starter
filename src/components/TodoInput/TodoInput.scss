
@import '../../globalStyles/mixins';

$greyText: rgba($primaryDarkText, 0.5);

.todoInput {
  @include itemBox();
  display: flex;
  position: relative;

  .title {
    flex-grow: 1;
    z-index: 1;
    min-width: 0;
  }

  .titleDefault {
    padding: inherit;
    position: absolute;
    top: 0;
    left: 0;
    color: $greyText;
  }

  .submitButton {
    height: 0;
    background: none;
    border: none;
    color: inherit;
    font-size: inherit;
    outline: none;
    padding: 0;
    padding-left: 1.4em;
    margin-top: -0.5em;

    .icon, .icon.disabled {
      padding-left: 0;
    }
  }

  .icon {
    @include icon(inherit, $hoverColor);
    
    &.disabled {
      @include icon($greyText, $greyText);
    }
  }
}
