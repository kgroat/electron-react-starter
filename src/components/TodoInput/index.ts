
export { default, State, Props } from './TodoInput'
