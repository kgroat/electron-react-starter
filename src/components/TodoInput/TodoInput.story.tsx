
import * as React from 'react'

import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import TodoInput from './'

storiesOf('TodoInput', module)
  .add('interactible', () => <TodoInput onAdd={action('added')} />)
