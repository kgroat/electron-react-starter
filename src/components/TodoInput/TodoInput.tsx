
import * as React from 'react'

import { todoInput, title, titleDefault, submitButton, icon, disabled } from './TodoInput.scss'

export interface State {
  name: string
  checked: boolean
}

export interface Props {
  onAdd: (name: string, checked: boolean) => void
}

export default class TodoInput extends React.Component<Props, State> {
  state = {
    name: '',
    checked: false
  }

  setName = (name: string) =>
    this.setState({ name })

  toggleChecked = (checked = !this.state.checked) =>
    this.setState({ checked })

  onAdd = (ev) => {
    ev.preventDefault()
    if (!this.state.name) {
      return
    }
    this.props.onAdd(this.state.name, this.state.checked)
    this.setState({ name: '', checked: false })
  }

  onNameChange = (ev: React.ChangeEvent<HTMLInputElement>) =>
    this.setName(ev.currentTarget.value)

  render () {
    const checkboxIcon = this.state.checked ? 'checkbox-marked-outline' : 'checkbox-blank-outline'

    return (
      <form className={todoInput} onSubmit={this.onAdd}>
        <input className={title} onChange={this.onNameChange} value={this.state.name} />
        { !this.state.name ? <span className={titleDefault}>Choose Name</span> : '' }
        <i onClick={() => this.toggleChecked()} className={`checkbox mdi mdi-${checkboxIcon} ${icon}`} />
        <button type='submit' className={submitButton}>
          <i className={`mdi mdi-plus-box ${icon} ${!this.state.name ? disabled : ''}`} />
        </button>
      </form>
    )
  }
}
