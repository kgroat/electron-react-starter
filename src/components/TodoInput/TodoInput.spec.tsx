
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import TodoInput from './'
import { title } from './TodoInput.scss'

describe('<TodoInput />', () => {
  describe('snapshots', () => {
    it('should render', () => {
      const tree = create(
        <TodoInput onAdd={jest.fn()} />
      )
      expect(tree).toMatchSnapshot()
    })
  })

  describe('user submitting on todo', () => {
    it('should not call onAdd if no name has been chosen', () => {
      const onAdd = jest.fn()
      const element = shallow(
        <TodoInput onAdd={onAdd} />
      )
      element.findWhere(el => el.is('form')).simulate('submit', { preventDefault: jest.fn() })
      expect(onAdd).not.toHaveBeenCalled()
    })
    it('should call onAdd with the name and false if a name has been chosen but the checkbox not checked', () => {
      const name = 'Get milk'
      const onAdd = jest.fn()
      const element = shallow(
        <TodoInput onAdd={onAdd} />
      )
      element.findWhere(el => el.hasClass(title)).simulate('change', {
        currentTarget: {
          value: name
        }
      })
      element.findWhere(el => el.is('form')).simulate('submit', { preventDefault: jest.fn() })
      expect(onAdd).toHaveBeenCalledWith(name, false)
      expect(onAdd).toHaveBeenCalledTimes(1)
    })
    it('should call onAdd with the name and true if a name has been chosen and the checkbox is checked', () => {
      const name = 'Get milk'
      const onAdd = jest.fn()
      const element = shallow(
        <TodoInput onAdd={onAdd} />
      )
      element.findWhere(el => el.hasClass(title)).simulate('change', {
        currentTarget: {
          value: name
        }
      })
      element.findWhere(el => el.hasClass('checkbox')).simulate('click')
      element.findWhere(el => el.is('form')).simulate('submit', { preventDefault: jest.fn() })
      expect(onAdd).toHaveBeenCalledWith(name, true)
      expect(onAdd).toHaveBeenCalledTimes(1)
    })
  })
})
