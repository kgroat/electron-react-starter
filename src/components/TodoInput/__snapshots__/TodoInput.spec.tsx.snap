// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoInput /> snapshots should render 1`] = `
<form
  className="todoInput"
  onSubmit={[Function]}
>
  <input
    className="title"
    onChange={[Function]}
    value=""
  />
  <span
    className="titleDefault"
  >
    Choose Name
  </span>
  <i
    className="checkbox mdi mdi-checkbox-blank-outline icon"
    onClick={[Function]}
  />
  <button
    className="submitButton"
    type="submit"
  >
    <i
      className="mdi mdi-plus-box icon disabled"
    />
  </button>
</form>
`;
